addon/-private/class/modifier.ts

```diff
@@ -64,6 +85,57 @@ export default class ClassBasedModifier<S = DefaultSignature> {
     deprecateForDestroyables('willDestroy', this);
     deprecateForDestroyables('isDestroying', this);
     deprecateForDestroyables('isDestroyed', this);
+
+    assert(
+      'ember-modifier: You cannot implement both `modify` and any of the deprecated legacy lifecycle hooks (`didInstall`, `didReceiveArguments`, and `didUpdateArguments`)',
+      !(_implementsModify(this) && _implementsLegacyHooks(this))
+    );
+  }
+
+  /**
+   * Called when the modifier is installed and any time any tracked state used
+   * in the modifier changes.
+   *
+   * If you need to do first-time-only setup, create a class field representing
+   * the initialization state and check it when running the hook. That is also
+   * where and when you should use `registerDestructor` for any teardown you
+   * need to do. For example:
+   *
+   * ```js
+   * function disconnect(instance) {
+   *  instance.observer?.disconnect();
+   * }
+   *
+   * class IntersectionObserver extends Modifier {
+   *   observer;
+   *
+   *   modify(element, callback, options) {
+   *     // Gets rid of any existing connection, since we're getting called with
+   *     // a new callback or new options, and drops any existing destructor
+   *     // registrations so we avoid having extra destructors lying around.
+   *     disconnect(this);
+   *     unregisterDestructor(this, disconnect);
+   *
+   *     this.observer = new IntersectionObserver(callback, options);
+   *     this.observer.observe(element);
+   *
+   *     registerDestructor(this, disconnect);
```

Comment:
Now that I read this again with fresh eyes, I think we _probably_ don't have to keep unregistering/reregistering since the actual thing we are doing **in this case** is using instance state to actually handle the disconnect (`instance.observer.disconnect()`) and the actual function is the same every time.

You would still need to either:

* unregister + reregister
* only register once

And which of those options that you choose probably depends on how your destructor works (e.g. does it use a stable instance prop, is it some other thing, etc).

---


FWIW, I think I'd personally do this for the case in the docs here (which only needs registration once, since the destructor is stable):

```js
function disconnect(instance) {
  instance.observer?.disconnect();
}

class IntersectionObserver extends Modifier {
  observer;

  constructor(owner, args) {
    super(owner, args);

    registerDestructor(this, disconnect);
  }

  modify(element, callback, options) {
    disconnect(this);
    this.observer = new IntersectionObserver(callback, options);
    this.observer.observe(element);
  }
}
```

Reply:
Ah, this seems better. Thank you!